src/ProductUsageService.php

```diff
@@ -0,0 +1,251 @@
+   *   Entity to update the mapping for.
+   */
+  public function update(EntityInterface $entity) {
+    if (!$entity instanceof FieldableEntityInterface) {
+      return;
+    }
+
+    if ($entity->getEntityTypeId() === 'paragraph') {
+      return;
+    }
+
+    // MenuLinkEntity returns NULL on call to $entity->id() so we check here to
+    // be sure.
+    if ($entity->id() === NULL) {
+      return;
+    }
+
+    $this->purge($entity->id(), $entity->getEntityTypeId());
+
+    $productFields = $this->fieldManager->getFieldMapByFieldType('amazon_product_widget_field_type');
+    $entityFields = $entity->getFieldDefinitions();
+
+    $hookAsins = $this->moduleHandler->invokeAll('amazon_product_widget_alter_asin_map', [$entity]);
+    $asins = [];
+    foreach ($entityFields as $fieldName => $definition) {
+      $targetType = $definition->getFieldStorageDefinition()->getSetting('target_type');
+      if (FALSE === array_key_exists($targetType, $productFields)) {
+        continue;
+      }
+
+      if ($definition->getType() === 'amazon_product_widget_field_type') {
+        try {
+          /** @var \Drupal\amazon_product_widget\Plugin\Field\FieldType\AmazonProductField $productField */
+          $productField = $entity->get($fieldName)->first();
+          if (!$productField instanceof AmazonProductField) {
+            continue;
+          }
+
+          $asins = array_merge($asins, $productField->getAsins());
+        }
+        catch (\Exception $exception) {
+          watchdog_exception('amazon_product_widget', $exception);
+        }
+      }
+      elseif ($definition->getType() === 'entity_reference_revisions') {
+        /** @var \Drupal\paragraphs\ParagraphInterface $paragraphField */
+        $paragraphField = $entity->get($fieldName);
+        if (!$paragraphField) {
+          continue;
+        }
+        $paragraphs = $paragraphField->referencedEntities();
+
+        /** @var \Drupal\paragraphs\ParagraphInterface $paragraph */
+        foreach ($paragraphs as $paragraph) {
+          $fields = $paragraph->getFieldDefinitions();
+          foreach ($fields as $field) {
+            if ($field->getType() === 'amazon_product_widget_field_type') {
+              try {
+                /** @var \Drupal\amazon_product_widget\Plugin\Field\FieldType\AmazonProductField $productField */
+                $productField = $paragraph->get($field->getName())->first();
+                if (!$productField instanceof AmazonProductField) {
+                  continue;
+                }
+                $asins = array_merge($asins, $productField->getAsins());
+              }
+              catch (\Exception $exception) {
+                watchdog_exception('amazon_product_widget', $exception);
+              }
+            }
+          }
+        }
+      }
+    }
+
+    $asins = array_merge($asins, $hookAsins);
+    $asins = array_unique($asins);
+    $this->insert($entity->id(), $entity->getEntityTypeId(), $asins);
```

Comment:
so purge above this line and check before existing entries with the results here